# Final Project: Predicting Covid Positive From Vaccines

In [1]:
import numpy as np
import pandas as pd
from   category_encoders             import *
from   sklearn.compose               import *
from   sklearn.ensemble              import *
from   sklearn.linear_model          import *
from   sklearn.impute                import *
from   sklearn.metrics               import *
from   sklearn.pipeline              import *
from   sklearn.preprocessing         import *
from   sklearn.model_selection       import *
from   sklearn.feature_selection     import *
from   imblearn.over_sampling        import *
from   sklearn.neural_network        import MLPClassifier
from   sklearn.gaussian_process      import GaussianProcessClassifier
from   sklearn.neighbors             import *
from   sklearn.naive_bayes           import *

## 1. Ask: Research Question / Hypothesis
----
#### **Research Question(s):** Given demographic information, which classification model most accurately predicts whether not someone will get Covid after getting the vaccine?  


#### **Hypothesis:** I believe the random forest classifier would provide the most accurate estimation for this classification problem since it's robust to outliers.


----

## 2. Acquire Data
----
#### **Data**: CDC's Vaccine Adverse Event Reporting System
1. Data Variables 
    - **id**: Unique Patient ID
    - **manu**: makers of vaccine products
    - **dose**: the number of vaccine doses given for a vaccine
    - **route**: route of administration/ the path the vaccine was taken into the body
    - **site**: the site for a vaccine product
    - **recovd_date:** indicates when the patient recovered from adverse symptoms
    - **state**: Patient's Home State
    - **age**: Patient's Age
    - **sex**: Patient's Sex
    - **hospdays**: Days spent in the hospital recovering from adverse symptoms
    - **disable**: Whether or not patient is disabled
    - **recvd**: Whether or not patient recovered from adverse symptoms
    - **numdays**: Number of days patient had adverse symptoms
    - **adminby**: public, private, other, military, work, pharmacy, senior living, school, unknown agency
    - **hosp_visit**: Whether or not patient had to visit the hospital from adverse symptoms
    - **er_visit**: Whether or not patient had to visit the ER from adverse symptoms
    - **history**: Patient's istory of diseases
    - **allergies**: Patient's allergies
    - **other_meds**: Patient's current medications
    - **symptom1** 1st adverse symptom from vaccine
    - **symptom2**: 2nd adverse symptom from vaccine
    - **symptom3**: 3rd adverse symptom from vaccine
    - **symptom4**: 4th adverse symptom from vaccine
    - **symptom5**: 5th adverse symptom from vaccine

In [78]:
df = pd.read_csv('vaccine_df.csv')

In [63]:
df

id             manu dose route state   age sex  hospdays disable  \
0       916600          MODERNA    1    IM    TX  33.0   F       NaN     NaN   
1       916601          MODERNA    1    IM    CA  73.0   F       NaN     NaN   
2       916602  PFIZER\BIONTECH    1    IM    WA  23.0   F       NaN     NaN   
3       916604          MODERNA    1    IM    TX  47.0   F       NaN     NaN   
4       916606          MODERNA    1    IM    NV  44.0   F       NaN     NaN   
...        ...              ...  ...   ...   ...   ...  ..       ...     ...   
18107  1057082  PFIZER\BIONTECH    1    IM    KY  86.0   M      12.0     NaN   
18108  1057082  PFIZER\BIONTECH    1    IM    KY  86.0   M      12.0     NaN   
18109  1057281          MODERNA    1    IM    MD  77.0   F       NaN     NaN   
18110  1057348  PFIZER\BIONTECH    1    IM    TX  88.0   F       NaN     NaN   
18111  1057795          MODERNA    1    IM    OH  82.0   F       NaN     NaN   

      recvd  ...                                         other_meds  \
0         Y  ...                                               None   
1         Y  ...  Patient residing at nursing facility. See pati...   
2         U  ...                                               None   
3         N  ...                                                 Na   
4         Y  ...                                                NaN   
...     ...  ...                                                ...   
18107     N  ...  Meds prior to admission: Amiodarone, ASA, Lipi...   
18108     N  ...  Meds prior to admission: Amiodarone, ASA, Lipi...   
18109     N  ...                                                NaN   
18110     N  ...                              lasik, blood pressure   
18111     N  ...                                                NaN   

                                                 history hosp_visit er_visit  \
0                                                   None          Y      NaN   
1      Patient residing at nursing facility. See pati...          Y      NaN   
2                                                   None        NaN        Y   
3                                                    NaN        NaN      NaN   
4                                                    NaN        NaN      NaN   
...                                                  ...        ...      ...   
18107  Arthritis, Afib, chronic anticoagulation, chro...        NaN      NaN   
18108  Arthritis, Afib, chronic anticoagulation, chro...        NaN      NaN   
18109                                                NaN        NaN      NaN   
18110                           congestive heart failure          Y      NaN   
18111                                                NaN        NaN      NaN   

                           allergies                  symptom1  \
0                  Pcn and bee venom                 Dysphagia   
1                            "Dairy"                   Anxiety   
2                          Shellfish          Chest discomfort   
3                                 Na   Injection site erythema   
4      iodine (shellfish) has epipen       Pharyngeal swelling   
...                              ...                       ...   
18107    Pneumococcal vaccine - rash                   Malaise   
18108    Pneumococcal vaccine - rash  SARS-CoV-2 test positive   
18109                            NaN                     Death   
18110                           none                     Death   
18111                            NaN                     Death   

                          symptom2                 symptom3  \
0                     Epiglottitis                      NaN   
1                         Dyspnoea                      NaN   
2                        Dysphagia        Pain in extremity   
3          Injection site pruritus  Injection site swelling   
4                              NaN                      NaN   
...                            ...                      ...   
181

## 3. Process Data
----

1. **Fill Specific NA Columns with Appropriate Values** 
    - hospdays: 0.0
    - disable: 'N'
    - hosp_visit: 'N'
    - er_visit: 'N'


2. **Merge the Symptoms into Singular Column** 
    - Create a new column that holds a flattened list of all symptoms 
    
    
3. **Create a new boolean column for every top allergy, other diagnosis, and medication**

   
4. **Create 'Covid' Target Column** 
    - There are two symptoms that encompass whether or not someone got covid 'SARS-CoV-2 test positive' and'COVID-19'. 
    - Need to merge those two columns together and return True if either of them are True
    
    
5. **Drop 'SARS-CoV-2 test positive', 'COVID-19' and 'All Symptoms' Columns**
    

In [64]:
class FillNaNs():
    def __init__(self, columns, values):
        self.columns = columns
        self.values = values
    
    def transform(self, X, **transform_params):        
        fillnas = {self.columns[i]: self.values[i] for i in range(len(self.columns))} 
        cpy_df = X.fillna(value=fillnas)
        return cpy_df

    def fit(self, X, y=None, **fit_params):
        return self
    
class MergeSymptomDFs():
    """Specialized class that inputs a list of symptom columns 
       and returns a new column with a flattened list of all the symptoms"""
    def __init__(self, columns, col, final_cols, final_cols2):
        self.columns = columns
        self.col = col
        self.final_cols = final_cols
        self.final_cols2 = final_cols2

    def transform(self, X, **transform_params): 
        cpy_df = X.copy()
        cpy_df2 = X.copy()
        
        # Groups df based on ID and joins symptoms together
        for symp in self.columns:
            cpy_df = pd.merge(cpy_df, cpy_df.groupby(self.col)[symp].apply(np.array).reset_index(name=symp), on=self.col)
        
        # Iterate through symptom columns and adds it all into one singular column
        for idx in range(len(self.columns)-1):
            cpy_df[self.columns[idx+1]+'_x'] = cpy_df[self.columns[idx]+'_x'] + ', ' + cpy_df[self.columns[idx+1]+'_x']
            cpy_df[self.final_cols[0]] = cpy_df[self.columns[idx]+'_x']
            cpy_df = cpy_df.drop(columns=[self.columns[idx]+'_x',self.columns[idx]+'_y'])
        
        # Joins cleaned symptom df to original df
        cpy_df = pd.merge(cpy_df[self.final_cols],cpy_df2[self.final_cols2], on = 'id')          
        return cpy_df

    def fit(self, X, y=None, **fit_params):
        return self  

    
class ConvertColumnLists():
    """Converts dataframe to have a new boolean column for every
       string in the list of strings 'list_strings' that returns
       whether or not the string is in 'column' """
    def __init__(self, list_strings, column):
        self.list_strings = list_strings
        self.column = column

    
    def transform(self, X, **transform_params): 
        cpy_df = X.copy()
        
        #creates a new column for every top adverse symptom
        for c in self.list_strings:
            cpy_df[c] = cpy_df[self.column].str.contains(c, case=False)
        return cpy_df

    def fit(self, X, y=None, **fit_params):
        return self

class BooleanInAnyDF():
    """ Inputs columns in a df 'cols', and returns a new column 
        'new_col' that indicates whether or not 'value' is in any 
        of the input columns, then drops the input columns"""
    def __init__(self, cols, value, new_col):
        self.cols = cols
        self.value = value
        self.new_col = new_col
    
    def transform(self, X, **transform_params): 
        cpy_df = X.copy()
        df = (cpy_df[self.cols].isin([self.value]))
        cpy_df[self.new_col] = (df.any(axis=1))
        return cpy_df

    def fit(self, X, y=None, **fit_params):
        return self

class DropColumns():
    def __init__(self, columns):
        self.columns = columns
    
    def transform(self, X, **transform_params):
        cpy_df = X.copy().drop(columns=self.columns)
        return cpy_df

    def fit(self, X, y=None, **fit_params):
        return self

In [101]:
fillNa = {'hospdays':0.0, 'disable':'N', 'hosp_visit':'N', 'er_visit':'N', 'history': 'None',
          'allergies': 'None', 'other_meds': 'None'}

symptoms = [f'symptom{i}' for i in range(1,6)]

final_cols = ['all_symptoms', 'id', 'manu', 'dose' ,'route', 'state', 'age', 'sex', 'hospdays', 'disable', 
              'history', 'numdays', 'adminby', 'hosp_visit','er_visit', 'other_meds', 'allergies']

covid_cols = ['SARS-CoV-2 test positive','COVID-19', 'all_symptoms', 'COVID']

text = ['history', 'allergies', 'other_meds', 'id']

history_list = ['asthma', 'arthritis', 'diabetes', 'migraines', 'hypothyroidism','gerd', 'high blood pressure', 'depression']

meds = ['levothyroxine', 'multivitamin', 'tylenol', 'adderall', 'ibuprofen', 'prenatal', 'levothyroxine', 'naltrexone']

allergies = ['penicillin', 'codeine', 'amoxicillin', 'shellfish', 'aspirin', 'lactose']

In [102]:
pipe_symp = Pipeline([('fillNaNs',            FillNaNs(list(fillNa.keys()), list(fillNa.values()))),             # Fill specific NA columns 
                      ('Merge Symptoms',      MergeSymptomDFs(symptoms, 'id', final_cols[:2], final_cols[1:])),  # Merge the symptoms into singular column
                      ('Expand Symptoms',     ConvertColumnLists(covid_cols[:2], final_cols[0])),                # Create a new boolean column for covid-19/cov test pos
                      ('Expand History',      ConvertColumnLists(history_list, 'history')),                      # Create a new boolean column for every top symptom
                      ('Expand Meds',         ConvertColumnLists(meds, 'other_meds')),                           # Create a new boolean column for every top medication
                      ('Expand Allerigies',   ConvertColumnLists(meds, 'allergies')),                            # Create a new boolean column for every top allergy
                      ('Create Target',       BooleanInAnyDF(covid_cols[:2], True, covid_cols[-1])),             # Create 'covid' target column
                      ('Drop Columns',        DropColumns(covid_cols[:3] + text)),                               # Drop unnecessary columns
                     ])

df = pipe_symp.fit_transform(df)

In [103]:
df

manu dose route state   age sex  hospdays disable  numdays  \
0              MODERNA    1    IM    TX  33.0   F       0.0       N      2.0   
1              MODERNA    1    IM    CA  73.0   F       0.0       N      0.0   
2      PFIZER\BIONTECH    1    IM    WA  23.0   F       0.0       N      0.0   
3              MODERNA    1    IM    TX  47.0   F       0.0       N      7.0   
4              MODERNA    1    IM    NV  44.0   F       0.0       N      0.0   
...                ...  ...   ...   ...   ...  ..       ...     ...      ...   
38157  PFIZER\BIONTECH    1    IM    KY  86.0   M      12.0       N      9.0   
38158  PFIZER\BIONTECH    1    IM    KY  86.0   M      12.0       N      9.0   
38159          MODERNA    1    IM    MD  77.0   F       0.0       N     14.0   
38160  PFIZER\BIONTECH    1    IM    TX  88.0   F       0.0       N      2.0   
38161          MODERNA    1    IM    OH  82.0   F       0.0       N     33.0   

      adminby  ... high blood pressure depression  levothyroxine  \
0         PVT  ...               False      False          False   
1         SEN  ...               False      False          False   
2         SEN  ...               False      False          False   
3         PUB  ...               False      False          False   
4         PVT  ...               False      False          False   
...       ...  ...                 ...        ...            ...   
38157     PVT  ...               False      False          False   
38158     PVT  ...               False      False          False   
38159     OTH  ...               False      False          False   
38160     PVT  ...               False      False          False   
38161     PUB  ...               False      False          False   

       multivitamin  tylenol  adderall  ibuprofen  prenatal  naltrexone  COVID  
0             False    False     False      False     False       False  False  
1             False    False     False      False     False       False  False  
2             False    False     False      False     False       False  False  
3             False    False     False      False     False       False  False  
4             False    False     False      False     False       False  False  
...             ...      ...       ...        ...       ...         ...    ...  
38157         False    False     False      False     False       False   True  
38158         False    False     False      False     False       False   True  
38159         False    False     False      False     False       False  False  
38160         False    False     False      False     False       False  False  
38161         False    False     False      False     False       False  False  

[38162 rows x 28 columns]

In [141]:
X = df.loc[:, (df.columns != 'COVID')]
y = df['COVID']

In [142]:
y.value_counts(normalize=True)

False    0.982103
True     0.017897
Name: COVID, dtype: float64

In [143]:
con_pipe = Pipeline([('imputer', SimpleImputer(strategy='median', add_indicator=True)), # replaces the NaN values with median
                     ('scaler', StandardScaler()),                                      # standardize the numerical features
                     ('var threshold', VarianceThreshold(0.06))]) 

cat_pipe = Pipeline([('imputer', SimpleImputer(strategy='most_frequent')), # replaces the NaN values with most frequent
                     ('encoder', OneHotEncoder(sparse=False)),             # represent categorical variables as binary vectors
                     ('var threshold', VarianceThreshold(0.06))]) 

final_prep_pipe = ColumnTransformer([('categorical', cat_pipe,   (X.dtypes == object)), 
                                    ('continuous',  con_pipe, ~(X.dtypes == object))])


In [144]:
X = final_prep_pipe.fit_transform(X)
y=y.values

### Ensure the Same Ratio of Classes for Train, Val and Test

In [145]:
stratSplit = StratifiedShuffleSplit(n_splits=1, test_size=0.2, random_state=42) 

for train_idx, test_idx in stratSplit.split(X, y): #ensures same underlying class distribution for train/test df
    X_train=X[train_idx]
    y_train=y[train_idx]
    X_test=X[test_idx]
    y_test=y[test_idx]

for train_idx, val_idx in stratSplit.split(X_train, y_train): #ensures same underlying class distribution for train/val df
    X_val= X_train[val_idx]
    y_val= y_train[val_idx]
    X_train=X_train[train_idx]
    y_train=y_train[train_idx]

## 4. Model 
----
#### **Balanced Accuracy:** Since we have imbalanced data, balanced accuracy score is best to account for both covid positive and not covid positive outcome classes.

### Without Oversampling

In [146]:
estimators = [('KNeighbors            ', KNeighborsClassifier()),
              ('RF Classifier         ', RandomForestClassifier()),
              ('SGD Classifier        ', SGDClassifier()),
              ('Extra Trees Classifier', ExtraTreesClassifier()),
              ('AdaBoost Classifier   ', AdaBoostClassifier()),
              ('MLP Classifier        ', MLPClassifier(max_iter=500)),
              ('Logistic Classifier   ', LogisticRegression(max_iter=1000))]

for estimator in estimators:
    pipe = Pipeline([('Classifier', estimator[1])])
    pipe.fit(X_train, y_train)
    y_pred = pipe.predict(X_val)
    print(f"{estimator[0]}: {round(balanced_accuracy_score(y_val, y_pred),3)}\n")


KNeighbors            : 0.548

RF Classifier         : 0.576

SGD Classifier        : 0.5

Extra Trees Classifier: 0.557

AdaBoost Classifier   : 0.5

MLP Classifier        : 0.561

Logistic Classifier   : 0.5



### SMOTE Oversampling

In [147]:
smote = SMOTE(sampling_strategy = 0.4) # duplicate instances of minority class
X_smote, y_smote = smote.fit_resample(X_train, y_train)

In [148]:
estimators = [('KNeighbors            ', KNeighborsClassifier()),
              ('RF Classifier         ', RandomForestClassifier()),
              ('SGD Classifier        ', SGDClassifier()),
              ('Extra Trees Classifier', ExtraTreesClassifier()),
              ('AdaBoost Classifier   ', AdaBoostClassifier()),
              ('MLP Classifier        ', MLPClassifier(max_iter=500)),
              ('Logistic Classifier   ', LogisticRegression(max_iter=1000))]

for estimator in estimators:
    pipe = Pipeline([('Classifier', estimator[1])])
    pipe.fit(X_smote, y_smote)
    y_pred = pipe.predict(X_val)
    print(f"{estimator[0]}: {round(balanced_accuracy_score(y_val, y_pred),3)}\n")

KNeighbors            : 0.638

RF Classifier         : 0.708

SGD Classifier        : 0.585

Extra Trees Classifier: 0.694

AdaBoost Classifier   : 0.606

MLP Classifier        : 0.734

Logistic Classifier   : 0.619



### Random Oversampling

In [149]:
oversample = RandomOverSampler(sampling_strategy = 0.4) # duplicate instances of minority class
X_oversample, y_oversample = oversample.fit_resample(X_train, y_train)

In [150]:
estimators = [('KNeighbors            ', KNeighborsClassifier()),
              ('RF Classifier         ', RandomForestClassifier()),
              ('SGD Classifier        ', SGDClassifier()),
              ('Extra Trees Classifier', ExtraTreesClassifier()),
              ('AdaBoost Classifier   ', AdaBoostClassifier()),
              ('MLP Classifier        ', MLPClassifier(max_iter=500)),
              ('Logistic Classifier   ', LogisticRegression(max_iter=1000))]

for estimator in estimators:
    pipe = Pipeline([('Classifier', estimator[1])])
    pipe.fit(X_oversample, y_oversample)
    y_pred = pipe.predict(X_val)
    print(f"{estimator[0]}: {round(balanced_accuracy_score(y_val, y_pred),3)}\n")


KNeighbors            : 0.684

RF Classifier         : 0.866

SGD Classifier        : 0.585

Extra Trees Classifier: 0.861

AdaBoost Classifier   : 0.669

MLP Classifier        : 0.839

Logistic Classifier   : 0.598



#### **Analysis 1:** Oversampling techniques improve model performance by providing more data from the minority class to learn from.
#### **Analysis 2:** Random Forest, Extra Trees, and MLP Classifiers performed the best.

In [151]:
estimators = {
    RandomForestClassifier: {
        'est__n_estimators': [100,300,500],
        'est__min_samples_leaf': [3,4,5]
    },
    ExtraTreesClassifier: {
        'est__n_estimators': [100,300,500],
        'est__min_samples_leaf': [3,4,5]
    },
    MLPClassifier: {
        'est__tol': [1e-3, 1e-4, 1e-5],
        'est__alpha': [1e-2, 1e-3, 1e-4],
        'est__max_iter':[1000]
    }
}

scores = []
for estimator, param_dist in estimators.items():

    pipe = Pipeline([('est', estimator())])

    rscv = RandomizedSearchCV(
        pipe,
        param_dist,
        n_jobs=-1,
    )
    rscv.fit(X, y)
    clean_best_param = {
        k.replace('est__', ''): v
        for k, v in rscv.best_params_.items()   
    }

    pipe = Pipeline([('est', estimator(**clean_best_param))])

    pipe.fit(X_oversample, y_oversample)
    y_pred = pipe.predict(X_val)
    score = balanced_accuracy_score(y_val, y_pred)
    scores.append((estimator.__name__, clean_best_param, score))

/Users/christabelle/anaconda3/lib/python3.8/site-packages/sklearn/model_selection/_search.py:278: UserWarning: The total space of parameters 9 is smaller than n_iter=10. Running 9 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(
/Users/christabelle/anaconda3/lib/python3.8/site-packages/sklearn/model_selection/_search.py:278: UserWarning: The total space of parameters 9 is smaller than n_iter=10. Running 9 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(
/Users/christabelle/anaconda3/lib/python3.8/site-packages/sklearn/model_selection/_search.py:278: UserWarning: The total space of parameters 9 is smaller than n_iter=10. Running 9 iterations. For exhaustive searches, use GridSearchCV.
  warnings.warn(


In [153]:
scores

[('RandomForestClassifier',
  {'n_estimators': 100, 'min_samples_leaf': 5},
  0.8470458470825627),
 ('ExtraTreesClassifier',
  {'n_estimators': 100, 'min_samples_leaf': 5},
  0.7933339758564297),
 ('MLPClassifier',
  {'tol': 0.001, 'max_iter': 1000, 'alpha': 0.01},
  0.8237826864502588)]

#### **Analysis:** Random Forest Performed Best

In [159]:
# additional hand hyperparamter tuning
pipe = Pipeline([('est', RandomForestClassifier(n_estimators=100,
                                                min_samples_leaf = 2,
                                                random_state=42))])

pipe.fit(X_oversample, y_oversample)
y_pred = pipe.predict(X_val)
print(f"RF Best Model: {round(balanced_accuracy_score(y_val, y_pred),3)}")

RF Best Model: 0.866


## 4. Deliver 
----
#### Test Set: Truest Measure of Generality

In [161]:
pipe = Pipeline([('est', RandomForestClassifier(n_estimators=100,
                                                min_samples_leaf = 2,
                                                random_state=42))])

pipe.fit(X_oversample, y_oversample)
y_pred = pipe.predict(X_test)
print(f"RF Best Model: {round(balanced_accuracy_score(y_test, y_pred),3)}\n") 

RF Best Model: 0.818



## Summary
1. With ~ 81.8% balanced accuracy, we can predict whether or not someone who receives a vaccine will still get covid
2. The best model turned out to be random forest, possibly due to the fact that it is robust to outliers and feature’s that aren’t as important will simply not get chosen during feature selection
3. Despite class imbalance, we were able to get the model to perform fairly well due to oversampling techniques.
4. This is a useful model for those who may feel uncertain about getting the vaccine and wants to know their chances of getting covid even after receiving the vaccine

## Limitations 
1. Model only used the top allergies, diseases, and medication. The data pipeline dropped all others that may have had higher predictive power.
2. A lot of other variables that would be predictive of whether or not someone still gets covid is missing (e.g, whether or not someone lives in a city).
3. Next steps: Continue working on feature engineering. Carry out methods to measure feature importance and encode more allergies/diseases/medication. 

